Feature/python benchmarking

vyasr opened the pull request:
<!--

Thank you for contributing to cuDF :)

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. There are CI checks in place to enforce that committed code follows our style
   and syntax standards. Please see our contribution guide in `CONTRIBUTING.MD`
   in the project root for more information about the checks we perform and how
   you can run them locally.

4. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

5. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please mark your pull request as Draft.
   https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request#converting-a-pull-request-to-a-draft

6. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove it from "Draft" and make it "Ready for Review".
   https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request#marking-a-pull-request-as-ready-for-review

   If assistance is required to complete the functionality, for example when the
   C/C++ code of a feature is complete but Python bindings are still required,
   then add the label `help wanted` so that others can triage and assist.
   The additional changes then can be implemented on top of the same PR.
   If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

7. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on the target branch, force push, or rewrite history.
   Doing any of these causes the context of any comments made by reviewers to be lost.
   If conflicts occur against the target branch they should be resolved by
   merging the target branch into the branch used for making the pull request.

8. Pull requests that modify cpp source that are marked ready for review 
   will automatically be assigned two cudf-cpp-codeowners reviewers.
   Ensure at least two approvals from cudf-cpp-codeowners before merging.

Many thanks in advance for your cooperation!

-->
This PR ports the benchmarks in https://github.com/vyasr/cudf_benchmarks, adding official benchmarks to the repository. The new benchmarks are designed from the ground up to make the best use of pytest, pytest-benchmark, and pytest-cases to simplify writing and maintaining benchmarks. Extended discussions of various previous design questions may be found on [the original repo](https://github.com/vyasr/cudf_benchmarks). Reviewers may also benefit from reviewing the companion PR creating documentation for how to write benchmarks, #11122.

Tests will not pass here until https://github.com/rapidsai/integration/pull/492 is merged.

vyasr commented:
This PR has a large changeset by necessity. In order to make this review process a bit smoother, I have explicitly requested reviews from people who have contributed benchmarks before and/or are familiar with my design for the benchmarks (@isVoid, @shwina, and @galipremsagar).

isVoid commented on python/cudf/benchmarks/conftest.py:
```diff
@@ -0,0 +1,232 @@
+for dtype, column_generator in column_generators.items():
+
+    def make_dataframe(nr, nc, column_generator=column_generator):
+        assert nc <= len(string.ascii_lowercase)
```
Let's put a comment here so that the assert error looks sensible when raised.

isVoid commented on python/cudf/benchmarks/common/utils.py:
Perhaps `make_fixture_and_inject_to_globals` is more descriptive?

vyasr replied:
It is more descriptive, but I think that's more detail than is really needed. This function is only two lines, and at that point the function name is almost as long as just copy-pasting the body of the function :) If someone really needs to understand the globals component they can read it.

isVoid commented on python/cudf/benchmarks/common/utils.py:
Why is sorted needed here?

isVoid commented on python/cudf/benchmarks/common/utils.py:
```diff
@@ -0,0 +1,268 @@
+    len_gather_map = round(len_gather_map)
+    len_column = round(len_column)
+
+    rstate = cupy.random.RandomState(seed=0)
```
Should seed be a global configuration?

vyasr replied:
I guess the question is how would that global seed be used? Do we use the same seed for all functions? For example does `make_boolean_mask_column` also use the same seed?

vyasr replied:
I agree that this is mostly a concern for tests, not benchmarks, which is why I didn't think a whole lot about this yet. I think we can afford to punt on making this decision in this PR and then spend a bit of time investigating afterwards.

isVoid commented on python/cudf/benchmarks/common/utils.py:
```diff
@@ -0,0 +1,268 @@
+    return cudf.core.column.as_column(rstate.randint(0, 2, size).astype(bool))
+
+
+def flatten(xs):
+    for x in xs:
+        if not isinstance(x, str):
+            yield from x
+        else:
+            yield x
```
It looks like this is valid syntax:
```python
In [5]: def flatten(xs):
   ...:     for x in xs:
   ...:         (yield x) if isinstance(x, str) else (yield from x)
```

vyasr replied:
It is. More importantly though, `flatten` is no longer needed :upside_down_face: I changed the part of the code that was originally using it, and I can delete this altogether now.

isVoid reviewed (commented):
Partial review part 1. Incredible work!

ajschmidt8 commented:
rerun tests

vyasr commented on python/cudf/benchmarks/common/config.py:
```diff
@@ -0,0 +1,69 @@
+      cupy). This feature enables easy comparisons of benchmarks between cudf
+      and pandas. All common modules (cudf, cupy) should be imported from here
+      by benchmark modules to allow configuration if needed.
+    - Defining CUDF_BENCHMARKS_TEST_ONLY will set global configuration
```
That is a very good point. `DEBUG_ONLY` seems like a reasonable choice.

vyasr commented:
@ajschmidt8 no need to rerun tests here until we merge https://github.com/rapidsai/integration/pull/492

ajschmidt8 commented:
@vyasr, https://github.com/rapidsai/integration/pull/492 has been merged

ajschmidt8 reviewed (approved):
Approving `ops-codeowner` file changes

isVoid commented on python/cudf/benchmarks/common/utils.py:
```diff
@@ -0,0 +1,267 @@
+    fixture : str, default None
+        The name of the fixture as used in the decorated test. If None,
+        defaults to `cls.lower()` if cls is a string, otherwise
+        `cls.__name__.lower()`. Use of this name allows the decorated function
+        to masquerade as a pytest receiving a fixture while substituting the
+        real fixture (with a much longer name).
```
It's not immediately clear to me what this means. Can you give an example?

vyasr replied:
Basically it means that you can write
```
@accepts_cudf_fixture(cls="column", dtype="int", nulls=False)
def test_foo(column)
```
and pytest will actually see a function
```
def test_foo(column_dtype_int_nulls_False)
``
Does that answer your question?

isVoid replied:
I guess my question is more with why `name` argument is needed for this utility function?

isVoid replied:
I personally find it confusing to provide a way to alias your parameter, especially when the name of `cls` is unambiguous. The only usage of this is if we have very complex union in the future that even the name of `cls` becomes non-intelligible, then we can think of introducing the aliases.

I agree with your definition of the decorator in its current state. I think the debating point is whether we should take the decorator a step further, hiding more details from the developer. It's a good addition, but probably beyond what this PR has to offer.

vyasr replied:
Even if we have a very complex union the `cls` should never become unintelligible; that would only be possible if our class hierarchy become complex and we introduced weird classes like `IndexedFrameWithoutNulls` or something like that, which I don't think will ever happen. @shwina what do you think about removing support for the `name` parameter entirely?

vyasr replied:
> I agree with your definition of the decorator in its current state. I think the debating point is whether we should take the decorator a step further, hiding more details from the developer (with other decorators). It's a good addition, but probably beyond what this PR has to offer.

What do you mean by this? What other details would you want to hide? Are you suggesting something like `accepts_cudf_fixture_with_nulls`, i.e. putting more of the parameter information into decorators that wrap this one? I don't think that adds any value, but perhaps I'm misunderstanding your suggestion.

isVoid replied:
Apologies, you should read this paragraph in the context of the comments I posted in the other PR: https://github.com/rapidsai/cudf/pull/11122#discussion_r903866124, I meant specifically that the fixture that @shwina proposed is the advanced version that hides more detail.

vyasr replied:
OK, and in that scenario it's only "hiding" in the sense of the name of the decorator, right? The function signature is not changed at all.

vyasr replied:
Discussed offline, we are removing the name parameter but otherwise leaving the decorator as is (a single decorator with `cls` as a parameter rather than one decorator per `cls`).

vyasr commented on python/cudf/benchmarks/common/utils.py:
```diff
@@ -0,0 +1,267 @@
+        exec(src, globals_)
+        wrapped_bm = globals_["wrapped_bm"]
+        # In case marks were applied to the original benchmark, copy them over.
+        if marks := getattr(bm, "pytestmark", None):
```
Nope. If you do
```
@accepts_cudf_benchmark(...)
@pytest.mark.parametrize(...)
def test_foo(...):
```
then marks are copied from `test_foo` onto the wrapper. If you swap that and do
```
@pytest.mark.parametrize(...)
@accepts_cudf_benchmark(...)
def test_foo(...):
```
then the marks are defined directly on the wrapper. In both cases, the wrapper gets all the marks.

isVoid replied:
Ah I see, didn't think of the other way.

vyasr commented on python/cudf/benchmarks/common/utils.py:
```diff
@@ -0,0 +1,267 @@
+        cls = cls.__name__
+    cls = cls.lower()
+
+    supported_classes = (
+        "column",
+        "series",
+        "index",
+        "dataframe",
+        "indexedframe",
+        "frame_or_index",
+    )
```
RangeIndex is not nullable and can't vary in dtype. In theory we could use the decorator for just sizes, I guess, but that seemed unnecessary. What do you think?

isVoid replied:
If I understand it correctly, including rangeindex here would change:
```python
def test_f(benchmark, rangeindex):
...
```
to
```python
@accept_cudf_fixture("rangeindex", size=[...])
def test_f(benchmark, rangeindex):
...
```
I don't think parameterizing the sizes is particularly useful for rangeindex as the default fixtures should suffice most use cases. I agree that we should leave it as a separate fixture union to avoid the complexity.

vyasr replied:
Yes, you understand correctly and we're in agreement that it doesn't seem very useful.

bdice commented on python/cudf/benchmarks/common/utils.py:
```diff
@@ -0,0 +1,267 @@
+                collects the last element
+    - random:   create a pseudorandom gather map
+
+    `len_gather_map`, `len_column` gets rounded to integer.
```
Why rounding instead of truncation, or requiring that inputs are `int` and not `Real`?

vyasr replied:
That's a question for @isVoid.

isVoid replied:
I think requiring inputs are `int` is the way to go here. @vyasr I don't think you should address this in this PR. I'll make an issue after you merge this.

bdice commented on python/cudf/benchmarks/common/utils.py:
```diff
@@ -0,0 +1,267 @@
+    )
+    assert cls in supported_classes, (
+        f"cls {cls} is invalid, choose from "
+        f"{', '.join(c for c in supported_classes)}"
```
```suggestion
        f"{', '.join(supported_classes)}"
```

bdice commented on python/cudf/benchmarks/common/utils.py:
```diff
@@ -0,0 +1,267 @@
+        dtype = [dtype]
+    assert all(dt in column_generators for dt in dtype), (
+        f"The only supported dtypes are "
+        f"{', '.join(dt for dt in column_generators)}"
```
```suggestion
        f"{', '.join(column_generators)}"
```

bdice commented on python/cudf/benchmarks/common/utils.py:
```diff
@@ -0,0 +1,267 @@
+    if cols is not None:
+        assert cols in NUM_COLS, (
+            f"You have requested a DataFrame with {cols} columns but fixtures "
+            f"only exist for the values {', '.join(c for c in NUM_COLS)}"
```
```suggestion
            f"only exist for the values {', '.join(NUM_COLS)}"
```

bdice commented on python/cudf/benchmarks/common/utils.py:
```diff
@@ -0,0 +1,267 @@
+    if rows is not None:
+        assert rows in NUM_ROWS, (
+            f"You have requested a {cls} with {rows} rows but fixtures "
+            f"only exist for the values {', '.join(c for c in NUM_ROWS)}"
```
```suggestion
            f"only exist for the values {', '.join(NUM_ROWS)}"
```

bdice commented on python/cudf/benchmarks/common/utils.py:
```diff
@@ -0,0 +1,267 @@
+    fixture_name = f"{cls}{dtype_str}{null_str}{col_str}{row_str}"
+
+    def deco(bm):
+        # pytests's test collection process relies on parsing the globals dict
```
```suggestion
        # pytest's test collection process relies on parsing the globals dict
```

bdice commented on python/cudf/benchmarks/common/utils.py:
Why not `*values`? Make it act more like `set`.
```suggestion
    def __init__(self, *values):
```

vyasr replied:
That isn't how set works
```
>>> set(1, 2, 3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: set expected at most 1 argument, got 3
>>> set((1, 2, 3))
{1, 2, 3}
>>> set([1, 2, 3])
{1, 2, 3}
```

bdice replied:
🤦‍♂️ Sorry, my bad.

bdice reviewed (commented):
This is great work. I have a lot of comments but I don't think any of them are major blockers. I'll hold off on approval until at least some of the suggestions are addressed, but I'd also be fine with merging this PR and fixing them later as well.

vyasr commented:
@bdice I will be addressing your comments, but broadly speaking I'd like to avoid making any further changes to individual benchmarks in this PR. A number of your questions are suggestions for improvements to benchmarks themselves, and while those are super worthwhile I'm going to punt on those. I'm going to focus on the changes that affect the overall benchmarks framework (the fixture setup, the various helper functions, etc). But feel free to continue making those other suggestions, I'll leave those threads open and address them in the subsequent PRs.

vyasr commented on python/cudf/benchmarks/API/bench_dataframe.py:
```diff
@@ -0,0 +1,115 @@
+
+@pytest.fixture(
+    params=[0, numpy.random.RandomState, cupy.random.RandomState],
+    ids=["Seed", "NumpyRandomState", "CupyRandomState"],
```
The performance does matter for a couple of reasons, mostly around the fact that our sampling implementation is different for sampling rows vs columns and you actually _can't_ use cupy arrays for column sampling (because it needs to happen on the host) but you could use either for sampling rows and there is a meaningful perf difference.

vyasr commented on python/cudf/benchmarks/common/config.py:
```diff
@@ -0,0 +1,69 @@
+
+def pytest_sessionfinish(session, exitstatus):
+    """Clean up sys.path after exit."""
+    if "common" in sys.path[0]:
```
You can do that, but what if we are using another plugin that also mucks with sys.path? I think the current solution of adding and then deleting is a minimally invasive solution that won't lead to breakage if another plugin also tries to do the same thing and doesn't include sufficient error checking.

vyasr replied:
I converted the benchmarks to a package, so now we can use relative imports and this issue is moot

vyasr commented:
@bdice @isVoid I think that I have addressed or responded to all of the comments, aside from those focusing on specific benchmarks that I intend to push to future work.

bdice commented on ci/gpu/build.sh:
```diff
@@ -235,6 +235,11 @@ cd "$WORKSPACE/python/cudf/cudf"
 gpuci_logger "Python py.test for cuDF"
 py.test -n 8 --cache-clear --basetemp="$WORKSPACE/cudf-cuda-tmp" --ignore="$WORKSPACE/python/cudf/cudf/benchmarks" --junitxml="$WORKSPACE/junit-cudf.xml" -v --cov-config="$WORKSPACE/python/cudf/.coveragerc" --cov=cudf --cov-report=xml:"$WORKSPACE/python/cudf/cudf-coverage.xml" --cov-report term --dist=loadscope tests
 
+# Run benchmarks with both cudf and pandas to ensure compatibility is maintained.
```
I'd put this after normal pytests and before notebook tests -- unless there's a reason that we can fail earlier with good error messages by running this before standard pytests.

I would also add in the comment that this only runs a _tiny_ benchmark input and not a full set of benchmarks.

bdice commented on python/cudf/benchmarks/API/bench_multiindex.py:
```diff
@@ -0,0 +1,43 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.
+
+import numpy as np
```
For consistency:
```suggestion
"""Benchmarks of MultiIndex methods."""

import numpy as np
```

bdice commented on python/cudf/benchmarks/internal/bench_column.py:
```diff
@@ -0,0 +1,116 @@
+#       column(int) (scatter)
+# value:    scalar,
+#           column (len(val) == len(key)),
+#           column (len(val) != len(key) & len == num_true)
```
```suggestion
#           column (len(val) != len(key) and len == num_true)
```

bdice reviewed (approved):
Really awesome job @vyasr. Excited to have these! I left a couple small comments but I'm +1 for punting on anything you want to address later.

galipremsagar commented on python/cudf/benchmarks/API/bench_dataframe.py:
```diff
@@ -0,0 +1,118 @@
+import numpy
+import pytest
+
+from ..common.config import cudf, cupy
+from ..common.utils import benchmark_with_object
```
Use absolute import?

galipremsagar commented on python/cudf/benchmarks/API/bench_frame_or_index.py:
```diff
@@ -0,0 +1,89 @@
+import numpy as np
+import pytest
+
+from ..common.utils import benchmark_with_object, make_gather_map
```
Here too

galipremsagar commented on python/cudf/benchmarks/API/bench_functions.py:
```diff
@@ -0,0 +1,53 @@
+import pytest
+import pytest_cases
+
+from ..common.config import cudf, cupy
```
Here too

galipremsagar commented on python/cudf/benchmarks/API/bench_functions_cases.py:
```diff
@@ -0,0 +1,149 @@
+
+import pytest_cases
+
+from ..common.config import NUM_ROWS, cudf, cupy
```
Here too

galipremsagar commented on python/cudf/benchmarks/API/bench_index.py:
```diff
@@ -0,0 +1,18 @@
+
+import pytest
+
+from ..common.config import cudf, cupy
+from ..common.utils import benchmark_with_object
```
here too

galipremsagar commented on python/cudf/benchmarks/API/bench_multiindex.py:
```diff
@@ -0,0 +1,45 @@
+import pandas as pd
+import pytest
+
+from ..common.config import cudf
```
here

galipremsagar commented on python/cudf/benchmarks/API/bench_series.py:
```diff
@@ -0,0 +1,24 @@
+
+import pytest
+
+from ..common.config import cudf, cupy
+from ..common.utils import benchmark_with_object
```
here too

galipremsagar commented on python/cudf/benchmarks/common/utils.py:
```diff
@@ -0,0 +1,258 @@
+
+import pytest_cases
+
+from .config import NUM_COLS, NUM_ROWS, cudf, cupy
```
Here

galipremsagar reviewed (commented):
For consistency sake can we switch all the relative imports to absolute imports?

vyasr commented:
> For consistency sake can we switch all the relative imports to absolute imports?

Done

mroeschke reviewed (approved):
Looks pretty good! This is a good base of benchmarks that can be expanded on in the future.

Not sure if this was discussed, but how did you imagine these benchmarks to be used? For example in pandas, we have used ASV to catch performance regressions across verions, but we're inconsistent about this as it takes a while to run the whole suite.

vyasr commented:
rerun tests

vyasr commented:
> Looks pretty good! This is a good base of benchmarks that can be expanded on in the future.
> 
> Not sure if this was discussed, but how did you imagine these benchmarks to be used? For example in pandas, we have used ASV to catch performance regressions across verions, but we're inconsistent about this as it takes a while to run the whole suite.

We have indeed discussed this before. ASV is something that we've definitely considered, not for running benchmarks but rather using the dashboards with data from pytest-benchmark via some plugins written by other RAPIDS team members. For the moment, the process of tracking historical benchmark results is pretty manual. We've discussed automating benchmark runs more going forward, but we haven't come to any specific solution yet. I think that to some extent this is waiting on various changes from our ops team.

vyasr commented:
rerun tests

vyasr commented on ci/gpu/build.sh:
```diff
@@ -243,6 +243,15 @@ cd "$WORKSPACE/python/custreamz"
+# Therefore, these runs only verify that benchmarks are valid.
+# They do not generate meaningful performance measurements.
+cd "$WORKSPACE"
+gpuci_logger "Python pytest for cuDF benchmarks"
```
```suggestion
gpuci_logger "Python pytest for cuDF benchmarks"
gpuci_logger "The path is ${PYTHONPATH}"
```

vyasr commented on ci/gpu/build.sh:
```diff
@@ -243,6 +243,15 @@ cd "$WORKSPACE/python/custreamz"
+# They do not generate meaningful performance measurements.
+cd "$WORKSPACE"
+gpuci_logger "Python pytest for cuDF benchmarks"
+CUDF_BENCHMARKS_DEBUG_ONLY=ON pytest -n 8 python/cudf/benchmarks
+CUDF_BENCHMARKS_USE_PANDAS=ON CUDF_BENCHMARKS_DEBUG_ONLY=ON pytest -n 8 python/cudf/benchmarks
```
```suggestion
CUDF_BENCHMARKS_DEBUG_ONLY=ON py.test -n 8 python/cudf/benchmarks
CUDF_BENCHMARKS_USE_PANDAS=ON CUDF_BENCHMARKS_DEBUG_ONLY=ON py.test -n 8 python/cudf/benchmarks
```

vyasr commented:
@bdice I had to roll back the changes making the benchmarks a Python package because it appears to change pytest's package discovery logic in a way that leads to it always finding the local copy of the package rather than the installed one, which is problematic when the package is not built in place (the case in our CI). I'm happy to revisit this further in the future, but in the interest of moving us forward I would like to merge as is and try to improve the organization and address the CI issues in future PRs.
